### DISCUSSION TITLE
useMutation for file uploading (mutation progress)

### DISCUSSION THREAD

**Discussion by nucleartux** [TanStack/query#1098] [General] [answered]
It's possible to make multi part request using useMutation and somehow get upload progress?
Any ideas?

**Comment by SimonKlausLudwig** [comment]
18 Thumbs Up and no answer ?

**Reply by cemcakirlar** [reply]
50 thumbs up now, I think I got the answer, see below :)

**Comment by TkDodo** [comment]
axios supports it, have you tried that?

https://gist.github.com/virolea/e1af9359fe071f24de3da3500ff0f429

Generally, this is likely out of scope for react-query, because it is agnostic towards the actual querying / mutating.

**Reply by nucleartux** [reply]
I don't think it is out of react-query scope, because basically I want to get mutation progress (nature of request doesn't matter actually, it can be upload progress or any other long time process).
So it will be very handy to get such information directly from `useMutation`, not only final result.

**Reply by TkDodo** [reply]
react-query doesn't know how you are going to do the actual upload. If you are using fetch, as far as I'm aware, there is no way to track upload progress if [this](https://stackoverflow.com/questions/35711724/upload-progress-indicators-for-fetch) is still accurate. Have you tried using local state alongside `onUploadProgress` from axios? This can be nicely tucked away in a custom hook (not sure if it works though). Maybe you can create a codesandbox example?

```
  const [completed, setCompleted] = React.useState(0);
  const uploadMutation = useMutation((formData) => {
    return axios.post("httpbin.org/post", formData, {
      onUploadProgress: function (progressEvent) {
        var percentCompleted = Math.round(
          (progressEvent.loaded * 100) / progressEvent.total
        );
        setCompleted(percentCompleted);
      }
    });
  });
```

**Reply by GabyMino** [reply]
Thank you! Your approach worked for our implementation. The only gotcha was that we needed to rethrow the axios error message if we wanted it to be available in the onError mutation side effect. I am not sure if that is the expected behavior

Component  (only relevant pieces included)
```
  const [completed, setCompleted] = useState(0);
  const { mutate: uploadFile, isSuccess: uploadSuccess } = useFileUploader();

  const upload = (file) => {
    uploadFile({ request: { file }, setCompleted });
  }

<LinearProgress className={classes.linearProgress} variant="determinate" value={completed} />

<div>
    <Button disabled={!acceptedFile} color="primary" onClick={() => upload(acceptedFile)}>
     Upload
    </Button>
</div>
```

Custom hook
```
export default function useFileUploader() {
  const snackbar = useSnackbar();
  return useMutation(
    async ({ request, setCompleted}: any) => {
      return await uploadFile({ request, setCompleted});
    },
    {
      onError: (e: any) => {
          snackbar.showError(e.message);
      },
      onSuccess: () => {
        snackbar.showSuccess('file uploaded successfully');
      },
    }
  );
}
```
Then in uploadFile function we rethrow

```
export const uploadFile = async({request, setCompleted}) => {
    const url = `my-url/file-upload`;

    const config = {
        onUploadProgress: function(progressEvent) {
          const percentCompleted = Math.round((progressEvent.loaded * 100) / progressEvent.total)
          setCompleted(percentCompleted);
        }
      }
    
      const data = new FormData();
      data.append('file', request.file);
    
      try{
        const res = await axios.post(url, data, config);
        return res;
      } catch(error){
        const errorMessage = error.response.data.message;
        throw new createHttpError(errorMessage);
      }
}
```

**Reply by TkDodo** [reply]
yes, if you catch an error, you need to re-throw it. react-query expects a failed promise returned in case of error, and async functions transform throw expressions to failed promises. if you catch the error, it will be a resolved promise, so this is expected :)

**Reply by brockklein** [reply]
Nice patterns everyone! fwiw here's our implementation. Short and sweet.

```typescript
import axios, { AxiosError } from 'axios'
import { useState } from 'react'
import { useMutation } from 'react-query'

interface IArgs {
    presignedUploadUrl: string
    file: File
}

export const useFileUploadMutation = () => {
    const [ progress, setProgress ] = useState(0)

    const mutation = useMutation<void, AxiosError, IArgs>(
        args => axios.put(
            args.presignedUploadUrl,
            args.file,
            { onUploadProgress: ev => setProgress(Math.round((ev.loaded * 100) / ev.total)) }
        )
    )

    return { ...mutation, progress }
}

```

**Reply by nikhil0o7** [reply]
is there a reason to use put instead of post?

**Reply by brockklein** [reply]
@nikhil0o7 I actually don't recall if there was a clear need for the PUT instead of POST. May be due to how we configured the presigned upload URL.

**Reply by LucasAndrad** [reply]
@nikhil0o7 if it's uploaded to S3 it's probably because S3 method is called putObject, but on the axios side it doesn't matter, POST, PUT and probably PATCH should work just fine.

**Reply by brockklein** [reply]
@LucasAndrad I bet you're right. I probably made it a `PUT` without thinking, just matching that naming from S3.

**Comment by cemcakirlar** [comment, marked as answer]
Old question but I want to share my way of solving the problem.

Because mutation function variables are agnostic enough (default type is any). I pass a callback function to track `ProgressEvent`  in `onUploadProgress`

I do post to express with uses multer. This example can handle single file per request.

```typescript
const post_path = '/api/upload';

export interface UseMutationUploadFileVariables {
    ParentLink_Category: string,
    ParentLink_RecID: string,
    RecId: string
    progressCallBack?: (progressEvent: ProgressEvent) => void
    file: File
}

const mutationFn_UploadFile = async (variables: UseMutationUploadFileVariables) => {

    const { ParentLink_Category, ParentLink_RecID, file, progressCallBack, RecId } = variables

    let fd = new FormData();

    //* order of appends matter, add file as last one, api endpoint is express with multer.
    fd.append("ParentLink_Category", ParentLink_Category)
    fd.append("ParentLink_RecID", ParentLink_RecID)
    fd.append("RecId", RecId)
    fd.append("uploaded_file", file)
    
    const axiosResponse = await axios.post<any, AxiosResponse<any>, FormData>(post_path, fd,
        {
            headers: {
                "Content-Type": "multipart/form-data",
                "Accept": "*/*"
            },
            onUploadProgress: (progressEvent: ProgressEvent) => {
                progressCallBack && progressCallBack(progressEvent)
            }
        });
    return axiosResponse
}

```

**Comment by thewasif** [comment]
If you are using **graphql-request** for posting GraphQL requests, then you do not need any extra code to upload file. You can upload with react-query `useMutation` hook without using form data or axios. Here, see an example 

```typescript
const UPDATE_IMAGE = gql`
    mutation updateImage($projectId: String!, $logoImage: Upload, $previewImage: Upload) {
        updateImage(projectId: $projectId, logoImage: $logoImage, previewImage: $previewImage) {
            previewImageUrl
            logoImageUrl
        }
    }
`;

const {mutate} = useMutation(async (params: any) => await request(GRAPHQL_BASE_URL, UPDATE_IMAGE, {...params}))

mutate({
   projectId: '1',
   projectImage: data.basics.projectImage, // JavaScript File
   logoImage: data.basics.logoImage // JavaScript File
})
 
```

You can refer to [graphql-request docs](https://github.com/prisma-labs/graphql-request#file-upload) to see more example

**Reply by PixeledCode** [reply]
`graphql-request ` has removed file upload feature

**Reply by thewasif** [reply]
Yes, unfortunately `graphql-request` has removed file upload feature. So you need to try something else or role back to older version which supports upload
